python313Packages.pyliblo: switch to pyliblo3 fork

ArchercatNEO opened the pull request:
Fixes #389149
Supersedes #418499

pyliblo has been unmaintained for 10 years and its required cpython version is incompatible with python 3.13.
A reputable fork exists which is compatible with pip and modern python tooling.
It is supported by Debian and raysession (one of the 2 apps which depends on pyliblo).


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Prince213 reviewed (changes requested):
Please add a update script and add yourself to maintainers.

ArchercatNEO committed:
maintainers: add archercatneo

Prince213 commented on maintainers/maintainer-list.nix:
Do this first so as not to break evaluation.

ArchercatNEO replied:
So add myself in the first commit and then in the second commit add the new package?

Prince213 commented on pkgs/development/python-modules/pyliblo3/default.nix:
```diff
@@ -0,0 +1,35 @@
+  src = fetchFromGitHub {
+    owner = "gesellkammer";
+    repo = "pyliblo3";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

Prince213 commented on pkgs/development/python-modules/pyliblo3/default.nix:
```diff
@@ -0,0 +1,35 @@
+  meta = {
+    homepage = "https://github.com/gesellkammer/pyliblo3/";
+    description = "Python wrapper for the liblo OSC library";
+    changelog = "https://github.com/gesellkammer/pyliblo3/blob/${src.rev}/NEWS";
```
```suggestion
    changelog = "https://github.com/gesellkammer/pyliblo3/blob/${src.tag}/NEWS";
```

Prince213 commented on pkgs/applications/audio/carla/default.nix:
```diff
@@ -36,6 +37,15 @@ stdenv.mkDerivation (finalAttrs: {
+    # Enable using the forked pyliblo3 over the unmaintained pyliblo
+    (fetchpatch2 {
+      name = "prefer-pyliblo3-over-pyliblo.patch";
+      url = "https://github.com/falkTX/Carla/pull/1933.patch";
```
Please read https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#fetching-patches.

ArchercatNEO replied:
Ah yes, PR urls are unstable and liable to be garbage collected, would this url be better `https://github.com/falkTX/Carla/pull/1933/commits/a81a2a545d2529233a6e0faa776fbd2d851442fb`? Or is there a different url which is stable?

ArchercatNEO replied:
I looked for usage of fetchpatch2 and got that for unmerged PRs this url style seems to work `https://github.com/falkTX/Carla/commit/a81a2a545d2529233a6e0faa776fbd2d851442fb.patch` (I got this from libretro same_cdi)

Prince213 replied:
You will still need ?full_index=1 but this may be more stable than the one using PR number?

dotlambda commented on pkgs/development/python-modules/pyliblo3/default.nix:
```diff
@@ -0,0 +1,35 @@
+buildPythonPackage rec {
+  pname = "pyliblo3";
+  version = "0.16.3";
+  format = "setuptools";
```
```suggestion
  pyproject = true;
```

dotlambda commented on pkgs/development/python-modules/pyliblo3/default.nix:
```diff
@@ -0,0 +1,35 @@
+    owner = "gesellkammer";
+    repo = "pyliblo3";
+    tag = "v${version}";
+    sha256 = "sha256-QfwZXkUT4U2Gfbv3rk0F/bze9hwJGn7H8t0X1SWqIuc=";
```
```suggestion
    hash = "sha256-QfwZXkUT4U2Gfbv3rk0F/bze9hwJGn7H8t0X1SWqIuc=";
```

dotlambda commented on pkgs/applications/audio/raysession/default.nix:
```diff
@@ -40,7 +40,7 @@ buildPythonApplication rec {
     bash
   ];
   propagatedBuildInputs = [
```
```suggestion
  dependencies = [
```

dotlambda commented on pkgs/development/python-modules/pyliblo3/default.nix:
```diff
@@ -0,0 +1,35 @@
+
+  buildInputs = [ liblo ];
+
+  passthru.updateScript = nix-update-script { };
```
```suggestion
```

ArchercatNEO replied:
I added this because the previous reviewer requested I add an update script. Is there a nixpkgs standard for whether there should/shouldn't be an update script?

dotlambda replied:
Python packages don't need it.

dotlambda commented on pkgs/development/python-modules/pyliblo3/default.nix:
`pythonImportsCheck` and `nativeCheckInputs` missing

ArchercatNEO replied:
Adding `pythonImportsCheck` works as expected (`pyliblo3` succeds, something like `fakepylib` fails) but I'm doubting what needs to go in `nativeCheckInputs` because I can't see any test dependencies or even tests being run in the build. Checking github actions reveals the test is just running python on the one file https://github.com/gesellkammer/pyliblo3/blob/484aa4b9dcbc15836796a724403d92f142a714de/.github/workflows/test.yml#L37 I suppose a custom checkPhase is needed to run this test?

dotlambda replied:
You can set `checkPhase` manually if `unittestCheckHook` or `pytestCheckHook` doesn't work.

dotlambda reviewed (commented):
Please make this separate commits:
* add pyliblo3
* use pyliblo3 in carla
* use pyliblo3 in raysession
* drop pyliblo and add a throw in `python-aliases.nix` pointing to pyliblo

ArchercatNEO committed:
python3Packages.pyliblo3: init at 0.16.3

This is a fork of pyliblo which has been unmaintained for 1 years.
It is reputable as Debian packages it and has support from at least
raysession.

ArchercatNEO committed:
carla: use pyliblo3

ArchercatNEO committed:
raysession: use pyliblo3

ArchercatNEO committed:
python3Packages.pyliblo: drop

Fixes #389149
Supersedes #418499

pyliblo has been unmaintained for 10 years and its required cpython version is incompatible with python 3.13.
Rather than keeping and patching a package with build issues we are
moving to a fork which builds under modern tooling.

timschumi commented on pkgs/top-level/python-aliases.nix:
```diff
@@ -558,6 +558,7 @@ mapAliases ({
   pyhiveapi = pyhive-integration; # Added 2025-02-22
   pyhs100 = throw "pyhs100 has been removed in favor of python-kasa."; # added 2024-01-05
   pylibgen = throw "pylibgen is unmaintained upstreamed, and removed from nixpkgs"; # added 2020-06-20
+  pyliblo = throw "pyliblo is unmaintained upstreamed and was removed from nixpkgs. Please use pyliblo3 instead"; # added 2025-06-23
```
```suggestion
  pyliblo = throw "pyliblo is unmaintained upstream and was removed from nixpkgs. Please use pyliblo3 instead"; # added 2025-06-23
```

timschumi commented on pkgs/development/python-modules/pyliblo3/default.nix:
I think you could init this package in `pkgs/by-name/` instead, would make future package updates a bit easier as well.

ArchercatNEO replied:
Yeah, seems like because python packages have a custom `callPackage` they aren't moving yet

https://github.com/NixOS/nixpkgs/blob/2ac2ae76dfd241b7c2e212c6200268e9f8237b63/pkgs/README.md?plain=1#L165-L167

timschumi replied:
It appears that the general replacement for those are `python3Packages.buildPythonPackage` and `python3Packages.cython` (there are many more python packages outside `by-name/py`), but yeah, maybe let's just keep it as-is for now.

timschumi reviewed (approved):
LGTM (but I'm not exactly an authority even when it comes to non-committer reviews)!

Also, please do mark feedback that has definitely been resolved (as in "both parties agree and the change, if necessary, has been carried out") as "resolved" in the UI, it makes the conversation slightly easier to scroll through.

Unfortunately, even after a significant number of years GitHub still hasn't managed to provide that button to the user that originally wrote the comment.

Prince213 reviewed (approved):
Diff LGTM

jys1670 commented:
Can we get this merged? At the very least, it makes Carla usable (tested), which is definitely an improvement over the completely broken state in the nixos-unstable channel right now.
![image](https://github.com/user-attachments/assets/a264cbc4-6d6a-4d70-b9f3-6bd30125265a)
